### ISSUE TITLE
RELEASE.2017-08-05T00-00-53Z breaks compatibility with Harbor registry

### ISSUE THREAD

**Issue by gregorkistler** [minio/minio#4779]
Hi,

first of all thanks for the instruction regarding how to use Harbor with Minio found [here](https://github.com/harshavardhana/harbor-minio) and [here](https://blog.minio.io/how-to-use-vmware-harbor-with-minio-c07a5c4ae31b).
Tried it yesterday but always ran into an issue while pushing an image. Actually it just retried the push action until it finally timed-out with message like 'received unexpected HTTP status: 500 Internal Server Error'.
Since the how-to is about 6 months old I thought it could be an issue with newer versions so I set up a test instance with Harbor 0.5.0 and the current Minio release which still failed. Then I downgraded Minio to version RELEASE.2017-02-16T01-47-30Z which was available back then and it suddenly worked.
So I tried every release up until the latest and greatest and figured out that only RELEASE.2017-08-05T00-00-53Z is affected, RELEASE.2017-07-24T18-27-35Z still does work like expected

The logs in Minio get spammed with:
```
time="2017-08-06T09:23:21Z" level=error msg="Received non-HTTP message from new connection 172.17.0.1:59372 at server 172.17.0.2:9000" cause="junk message" source="[listener.go:205:github.com/minio/minio/pkg/http.(*httpListener).start.func2()]" 
time="2017-08-06T09:23:21Z" level=error msg="Received non-HTTP message from new connection 172.17.0.1:59376 at server 172.17.0.2:9000" cause="junk message" source="[listener.go:205:github.com/minio/minio/pkg/http.(*httpListener).start.func2()]" 
time="2017-08-06T09:23:21Z" level=error msg="Received non-HTTP message from new connection 172.17.0.1:59380 at server 172.17.0.2:9000" cause="junk message" source="[listener.go:205:github.com/minio/minio/pkg/http.(*httpListener).start.func2()]"
```

Harbor registry container log shows a couple of those messages:
```
Aug  6 11:23:21 172.18.0.1 registry[10344]: time="2017-08-06T09:23:21.273836923Z" level=debug msg="s3aws.PutContent(\"/docker/registry/v2/repositories/test2/busybox/_uploads/7f2f68b6-71c2-4ebf-9c4b-ecf575c9c5ca/startedat\")" auth.user.name=admin go.version=go1.6.3 http.request.host=10.0.1.67 http.request.id=b5e7a450-c0bb-4015-a72e-edd5af5dc254 http.request.method=POST http.request.remoteaddr=10.0.1.98 http.request.uri="/v2/test2/busybox/blobs/uploads/" http.request.useragent="docker/17.06.0-ce go/go1.8.3 git-commit/02c1d87 kernel/4.9.36-moby os/linux arch/amd64 UpstreamClient(Docker-Client/17.06.0-ce \\(darwin\\))" instance.id=d59d2861-01f6-45d6-8c4f-4ba2150943be service=registry trace.duration=310.67379ms trace.file="/go/src/github.com/docker/distribution/registry/storage/driver/base/base.go" trace.func="github.com/docker/distribution/registry/storage/driver/base.(*Base).PutContent" trace.id=925c6a31-d956-493f-891b-6e6422d7d65a trace.line=95 vars.name="test2/busybox" version=v2.5.0 
Aug  6 11:23:21 172.18.0.1 registry[10344]: time="2017-08-06T09:23:21.274006542Z" level=error msg="response completed with error" auth.user.name=admin err.code=unknown err.detail="s3aws: RequestError: send request failed\ncaused by: Put https://10.0.1.67:9000/harbor-objects/docker/registry/v2/repositories/test2/busybox/_uploads/7f2f68b6-71c2-4ebf-9c4b-ecf575c9c5ca/startedat: EOF" err.message="unknown error" go.version=go1.6.3 http.request.host=10.0.1.67 http.request.id=b5e7a450-c0bb-4015-a72e-edd5af5dc254 http.request.method=POST http.request.remoteaddr=10.0.1.98 http.request.uri="/v2/test2/busybox/blobs/uploads/" http.request.useragent="docker/17.06.0-ce go/go1.8.3 git-commit/02c1d87 kernel/4.9.36-moby os/linux arch/amd64 UpstreamClient(Docker-Client/17.06.0-ce \\(darwin\\))" http.response.contenttype="application/json; charset=utf-8" http.response.duration=321.180246ms http.response.status=500 http.response.written=104 instance.id=d59d2861-01f6-45d6-8c4f-4ba2150943be service=registry vars.name="test2/busybox" version=v2.5.0 
Aug  6 11:23:21 172.18.0.1 registry[10344]: 172.18.0.7 - - [06/Aug/2017:09:23:20 +0000] "POST /v2/test2/busybox/blobs/uploads/ HTTP/1.1" 500 104 "" "docker/17.06.0-ce go/go1.8.3 git-commit/02c1d87 kernel/4.9.36-moby os/linux arch/amd64 UpstreamClient(Docker-Client/17.06.0-ce \\(darwin\\))"
```

## Expected Behavior
Harbor registry works flawlessly together with Minio as an image storage

## Current Behavior
Docker push keeps retrying for 30 to 60 seconds and finally fails with a 500 HTTP error

## Steps to Reproduce (for bugs)
1. Install Minio - Docker container or binary .. doesn´t matter
2. Install Harbor as described in the [official documentation](https://github.com/vmware/harbor/blob/master/docs/installation_guide.md) or as written in your own [README.md](https://github.com/harshavardhana/harbor-minio/blob/master/README.md)
3. Create a new project using the Harbor UI 
4. Tag an image
5. Log in to the Harbor registry via Docker
6. Push the tagged image

## Your Environment
* Version used (`minio version`):
**RELEASE.2017-07-24T18-27-35Z**
RELEASE.2017-06-13T19-01-01Z
RELEASE.2017-05-05T01-14-51Z
...
RELEASE.2017-02-16T01-47-30Z
* Operating System and version (`uname -a`):
Test environment was a single CentOS installation with both Minio and Harbor
`Linux centos-minio1 3.10.0-514.26.2.el7.x86_64 #1 SMP Tue Jul 4 15:04:05 UTC 2017 x86_64 x86_64 x86_64 GNU/Linux`
Prod environment runs Minio on CentOS and Harbor on VMware´s Photon OS - with the same behaviour.

**Comment by harshavardhana** [comment]
Are you using TLS @gregorkistler ?  The error looks like harbor is trying to do TLS connection and Minio server is rejecting it since we stopped supporting TLS and non-TLS (by redirecting to TLS) on the same connection due to security reasons.

**Comment by harshavardhana** [comment]
Can you also provide logs from minio console if any?

**Comment by gregorkistler** [comment]
Nope, haven´t configured Minio server to use TLS - at least not in my test env.
Prod env does SSL termination at a load balancer but also no encryption regarding Minio itself.

**Comment by krisis** [comment]
@gregorkistler I am facing difficulties in getting an Harbor setup up and running as described in their official documentation. I shall post my findings on this issue when I cross this.

**Comment by donatello** [comment]
@gregorkistler Apologies to come back to this after a long time, however, I have just verified that Harbor registry works with `RELEASE.2017-10-27T18-59-02Z` (latest stable release as of this writing). 

I used the instructions given in https://github.com/harshavardhana/harbor-minio/blob/master/README.md - and pushing an image to the Harbor registry works correctly.

N.B. Since my test setup was not meant for production, I deployed Harbor without [HTTPS](https://github.com/vmware/harbor/blob/master/docs/configure_https.md) - so when logging in to the registry with `docker login http://<harbor-server-ip>`, I had to make sure that the docker daemon considered the registry as insecure, using the `insecure-registries` option in the [daemon configuration file](https://docs.docker.com/engine/reference/commandline/dockerd//#daemon-configuration-file).
